fsdp/minimal-fsdp/fsdp.py

```diff
@@ -0,0 +1,339 @@
+import json
+import logging
+import os
+import random
+from typing import Any, Dict, Generator, Optional, TypedDict
+
+import determined as det
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+from torch.distributed.fsdp import FullStateDictConfig
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp import ShardingStrategy, StateDictType
+from torch.distributed.fsdp.sharded_grad_scaler import ShardedGradScaler
+from torch.distributed.fsdp.wrap import ModuleWrapPolicy
+
+from model import EmbedAndEncode, LMHead, Transformer, TransformerBlock
+
+"""
+Minimal transformer model FSDP script with Core API.
+"""
+
+
+def get_fake_data_iter(
+    batch_size: int,
+    vocab_size: int,
+    max_seq_len: int,
+    rank: int,
+    device: torch.device,
+    is_validation: bool,
+    simulated_size_in_batches: int = 10,
+) -> Generator[tuple[torch.Tensor, torch.Tensor], None, None]:
+    """
+    Fake dataloader. Yields a different set of data for each rank, and for train vs validation.
+    This data would usually come from a tokenized dataset.
+    """
+    generator = torch.Generator(device=device)
+    next_idx = 0
+    while True:
+        if next_idx == 0:
+            generator.manual_seed(42 + rank + 100000 * is_validation)
```

Comment:
Just curious: why this if block? Why not seed immediately after creating the generator?

Also, this isn't very important, but I guess the restarting logic starts the dataloader over, rather than continuing from the last batch. I wouldn't fix this, just maybe note it in a comment.

Reply:
Goal: I'm trying to have the generator yield a recurring sequence of length `simulated_size_in_batches`.  I do this by reseeding to the same initial seed after `simulated_size_in_batches` steps, which comes from setting `next_idx = (next_idx + 1) % simulated_size_in_batches`.

I save repeating the code by having the initial seed and the reseeds use the same if condition here.

Reply:
ah, I missed the modulo, got it now